packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -119,30 +119,36 @@ export function useFocusManager(): FocusManager {
 function createFocusManager(scopeRef: React.RefObject<HTMLElement[]>): FocusManager {
   return {
     focusNext(opts: FocusManagerOptions = {}) {
-      let node = opts.from || document.activeElement;
-      let focusable = getFocusableElementsInScope(scopeRef.current, opts);
-      let nextNode = focusable.find(n =>
-        !!(node.compareDocumentPosition(n) & (Node.DOCUMENT_POSITION_FOLLOWING | Node.DOCUMENT_POSITION_CONTAINED_BY))
-      );
-      if (!nextNode && opts.wrap) {
-        nextNode = focusable[0];
+      let scope = scopeRef.current;
+      let {from, tabbable, wrap} = opts; 
+      let node = from || document.activeElement;
+      let sentinel = scope[0].previousElementSibling;
+      let walker = getFocusableTreeWalker(getScopeRoot(scope), {tabbable});
+      walker.currentNode = isElementInScope(node, scope) ? node : sentinel;
+      let nextNode = walker.nextNode() as HTMLElement;
+      if ((!nextNode || !isElementInScope(nextNode, scope)) && wrap) {
+        walker.currentNode = sentinel;
+        nextNode = walker.nextNode() as HTMLElement;
       }
-      if (nextNode) {
-        nextNode.focus();
+      if (nextNode && isElementInScope(nextNode, scope)) {
```

Comment:
All the `isElementInScope` checks make me wonder if there should be a way to add that into the tree walker filter so we don't have to do this in so many places...

Reply:
I guess we have to distinguish between a tree walker that returns only elements in scope and one that includes the whole DOM, the `focusPrevious/focusNext` use case versus the `onKeyDown/Tab` use case.

Reply:
I've managed to reduce the number of calls to `isElementInScope` with: https://github.com/adobe/react-spectrum/pull/1493/commits/c63c243866630322c0315e0243ceae6ede32f472.